src/graph.cc

```diff
@@ -135,6 +135,7 @@ bool DependencyScan::RecomputeDirty(Node* node, vector<Node*>* stack,
       } else {
         if (!most_recent_input || (*i)->mtime() > most_recent_input->mtime()) {
           most_recent_input = *i;
+          edge->most_recent_input_mtime = (*i)->mtime();
```

Comment:
What happens when Edge is already part of `ready_` at this point?

Reply:
Nice question!

Per local experiments, it stays in the same order as it was. I.e. if I iterate over the map after I changed the first element to be last one, it stays the first. Worth noting though that upon asking for the element directly, it will not be found. Out of interest, I'll look if it's possible to re-sort a container — it looks like nobody have ever asked that question.

It doesn't matter though, because in ninja all of the work with mtime happens before adding the edges, so it's not affected.

But of course alternatives are welcome.

Reply:
My idea:

```cpp
struct EdgeHandle {
  Edge* const p;
  const TimeStamp mtime;
};
```

And use this for `ready_`. This way the compiler will make sure, that an element can't be changed after it has been added.

Still one could add an Edge that's already part of the set a second time ... hm ... also it would make accessing the set pretty cumbersome.

Okay forget the idea :D

Maybe we should just use `std::multiset` instead, so that it becomes apparent that uniqueness isn't guaranteed?

Reply:
> Maybe we should just use `std::multiset` instead, so that it becomes apparent that uniqueness isn't guaranteed?

Problem isn't lack of uniqueness, but that the container relies on the order to search for an element. So after mtime in an Edge was modified, the container needs to be reshuffled, otherwise `find()` method won't find an element. This holds for multiset as well.

Honestly, I think the only way to make this future-proof, is to rewrite a lot of ninja code to provide more immutability guaranteees. In particular to make it explicit that once `Plan` started gathering Nodes/Edges, they won't be modified.

I made a quick stab at constifying the Plan — I'm not sure yet if it's easy, but for sure it would be a lot of changes.

Reply:
Ah I see. Maybe just use a plain old std::vector and do the sorting on insert ourself?

Reply:
> Ah I see. Maybe just use a plain old std::vector and do the sorting on insert by ourselves?

Well, yeah, I can do it. I'd probably need to write a wrapper around `std::vector`, to make sure sorting is always done on `push_back`.

> I made a quick stab at constifying the Plan — I'm not sure yet if it's easy, but for sure it would be a lot of changes.

Yeah, definitely not easy. One would need to decouple various initialization code to make things immutable, just inserting `const` here and there unfortunately doesn't work.

Reply:
I asked on IRC #C++, turns out [keys to an associative container must be immutable](https://stackoverflow.com/a/15701171/2388257#why-not-modify-key-of-associative-container). Mutating them is undefined behavior, so they should ideally be const everywhere.

One of sad consequences of C++ not enforcing immutability by default, having it as an opt-in feature.

Reply:
@jhasse phew, that should be ready now. First commit adds the wrapper and migrates `ready_` there, the second adds mtime to edges and acc. comparisons.

Reply:
I think it is not good to use vector for sorted set.
It looks to cause O(n^2) time complexity.

Can we just remove edge from ready_ before modification and re-insert edge after that?

Reply:
> Can we just remove edge from ready_ before modification and re-insert edge after that?

I don't think they're modified after they got into the `ready_`. But this can't guarantee that someone not gonna add the code that does modify them.

Reply:
std::sort is O(n*log(n)) IIRC.